### ISSUE TITLE
kubernetes operator: delete of CRs hangs on finalizer - sometimes

### ISSUE THREAD

**Issue by axelsean** [hazelcast/hazelcast-platform-operator#651]
When deleting the CRs, from the kubernetes operator, the delete hangs on finalizers - around 25% of time.  This is in Azure.

**Background:**

Our application uses hazelcast to cache:
- maps
- java objects

When the application is upgraded there's a blue/green upgrade, scale down the blue, deploy green (test), delete blue (if all good).  On the scale down the CRs are deleted (not tried patching clusterSize to zero).  It's this delete that hangs.

The clusters in Azure usually have multiple namespaces defined, each namespace is a full application, thus the CRD is shared (since this doesn't seem to be namespaced).   Each namespace is upgraded separately.

The issue is not multiple versions of hazelcast - that's hard coded - see below.

There's a secondary use case (for our develoment builds) where we also delete the whole application and the namespace - same STR and frequency of failure.

**Expected behavior**
The CRs are deleted

**To Reproduce**

- delete the hazelcast CR
- delete the management CR
- (optional) delete the namespace

```bash
kubectl delete managementcenters.hazelcast.com/ams-hazelcast-managementcenter-blue -n my-namespace
kubectl delete Hazelcast/ams-hazelcast-blue  -n my-namespace
```
.

**Additional context**

We've observed 3 cases of issue:

- the hazelcast CR hangs and it has a finalizer
- the management CR hangs and it has a finalizer
- when deleting the namespace it hangs and we have to remove the finalizer

**Work Arounds**

For the CRs, we have patched to remove the finalizer, this 

```bash
kubectl patch Hazelcast/ams-hazelcast-blue -n my-namespace -p '{"metadata":{"finalizers":[]}}' --type=merge
kubectl patch managementcenters/ams-hazelcast-managementcenter-blue -n my-namespace -p '{"metadata":{"finalizers":[]}}' --type=merge

```

For the namespace issue
```bash
kubectl get namespace my-namespace -o json   | tr -d "\n" | sed "s/\"finalizers\": \[[^]]\+\]/\"finalizers\": []/"   | kubectl replace --raw /api/v1/namespaces/my-namespace/finalize -f -

```

One result we observed:

```bash
get namespace my-namespace -o json 
{
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"Namespace\",\"metadata\":{\"annotations\":{},\"name\":\"ams-hazelcast2-nonprod\"}}\n"
        },
        "creationTimestamp": "2023-03-02T17:30:43Z",
        "labels": {
            "kubernetes.io/metadata.name": "ams-hazelcast2-nonprod"
        },
        "name": "my-namespace",
        "resourceVersion": "3304666",
        "uid": "466da2a1-a4bf-497a-9b64-dfc9d082784f"
    },
    "spec": {
        "finalizers": [
            "kubernetes"
        ]
    },
    "status": {
        "phase": "Active"
    }
}
```

**CR definiitions**

setup:
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: 'my-namespace'
#
# See also https://docs.hazelcast.com/operator/latest/get-started
#
resources:
  - https://repository.hazelcast.com/operator/bundle-5.5.yaml
```
hazelcast CR
 
```yaml
apiVersion: hazelcast.com/v1alpha1
kind: Hazelcast
metadata:
  name: ams-hazelcast-@SLOT@
  namespace: '@STACK_NAMESPACE@'         
spec:                                    
  clusterSize: 3                         
  clusterName: ams-hazelcast-cluster-@SLOT@
  repository: 'docker.io/hazelcast/hazelcast'
  version: '5.1.4'
  exposeExternally:                      
    memberAccess: LoadBalancer      
    type: Smart                           
    discoveryServiceType: NodePort
  loggingLevel: INFO
  properties:                                   
    hazelcast.jmx: "true"
    hazelcast.shutdownhook.policy: TERMINATE      # Default - GRACEFUL waits for closedown - could be blocked on finalizer
    hazelcast.shutdownhook.enabled: "true"
    hazelcast.graceful.shutdown.max.wait: "600"
```

ManagementCenter CR

```yaml
apiVersion: hazelcast.com/v1alpha1
kind: ManagementCenter
metadata:
  name: ams-hazelcast-managementcenter-blue
  namespace: 'my-namespace'
spec:
  repository: 'hazelcast/management-center'
  version: '5.1.4'
  externalConnectivity:
    type: LoadBalancer
  hazelcastClusters:
    - address: ams-hazelcast-blue
      name: ams-hazelcast-cluster-blue
  persistence:
    enabled: true
    size: 10Gi
```

**Comment by mtyazici** [comment]
Hi @axelsean, I transfered the issue as https://github.com/hazelcast/hazelcast-platform-operator/issues/651. Closing the issue in this repo.

**Comment by axelsean** [comment]
@mtyazici closing appears to have closed it in the kubernetes/operator repo

**Comment by axelsean** [comment]
Once more interesting thing, when I tried to delete the HZ cluster, it wouldn't go into terminating state - just stayed running - though the command reported it was deleted.

```
 kubectl delete managementcenter ams-hazelcast-managementcenter-blue  -n my_namespace--wait=false
managementcenter.hazelcast.com "ams-hazelcast-managementcenter-blue" deleted

kubectl delete Hazelcast/ams-hazelcast-blue -n my_namespace

hazelcast.hazelcast.com "ams-hazelcast-blue" deleted

```

**Comment by cagric0** [comment]
I am currently trying to reproduce it. In the meantime could you share operator and hazelcast logs and also hazelcast cr yaml when it hangs?

**Comment by axelsean** [comment]
@cagric0 There are yaml for the CR above - both HZ and the ManagementCenter - note that we replace the @SLOT@ (with "blue") and the @STACK_NAMESPACE@ before deploy.

We've upgraded to 5.5 of hazelcast - still getting the issue sometimes

The logs requested - are you asking for the pod logs for (these are the pods we have running):

- hazelcast-platform-controller-manager-...
- ams-hazelcast-blue-0 (and 1 and 2)
- ams-hazelcast-managementcenter-blue-0

**Comment by cagric0** [comment]
Hi @axelsean, I've reproduced the issue and found the cause reason. It seems we've already fixed the issue with version **5.6**, when we change the operator deployment strategy from `bundle yaml` to `helm installation`. The problem was related to the cluster role assigned to the operator when we install a new operator for each namespace. Thus we have a proper version for your setup. I would suggest you upgrade your operator to version **5.8**(latest). You can check the installation instructions for two different use case:
1- Cluster Wide Installation
2- Restricted Installation
Here is the doc page: https://docs.hazelcast.com/operator/latest/get-started

**Comment by cagric0** [comment]
I am hope the new versions are working for you. I am closing this issue, but If you have encountered same or similar issue again, you can always open this issue or create new one.
